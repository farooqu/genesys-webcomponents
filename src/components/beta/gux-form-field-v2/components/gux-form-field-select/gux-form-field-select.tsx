import { Component, Element, h, JSX, Prop, State } from '@stencil/core';

import { OnMutation } from '../../../../../utils/decorator/on-mutation';
import { onDisabledChange } from '../../../../../utils/dom/on-attribute-change';
import { onRequiredChange } from '../../../../../utils/dom/on-attribute-change';
import { preventBrowserValidationStyling } from '../../../../../utils/dom/prevent-browser-validation-styling';
import { trackComponent } from '../../../../../usage-tracking';

import { GuxFormFieldContainer } from '../../functional-components/gux-form-field-container/gux-form-field-container';
import { GuxFormFieldError } from '../../functional-components/gux-form-field-error/gux-form-field-error';
import { GuxFormFieldLabel } from '../../functional-components/gux-form-field-label/gux-form-field-label';

import { GuxFormFieldLabelPosition } from '../../gux-form-field.types';
import {
  hasErrorSlot,
  getComputedLabelPosition,
  validateFormIds
} from '../../gux-form-field.servce';

/**
 * @slot input - Required slot for input tag
 * @slot label - Required slot for label tag
 * @slot error - Optional slot for error message
 */
@Component({
  styleUrl: 'gux-form-field-select.less',
  tag: 'gux-form-field-select-beta',
  shadow: true
})
export class GuxFormFieldSelect {
  private input: HTMLSelectElement;
  private label: HTMLLabelElement;
  private disabledObserver: MutationObserver;
  private requiredObserver: MutationObserver;

  @Element()
  private root: HTMLElement;

  @Prop()
  labelPosition: GuxFormFieldLabelPosition;

  @State()
  private computedLabelPosition: GuxFormFieldLabelPosition = 'above';

  @State()
  private disabled: boolean;

  @State()
  private required: boolean;

  @State()
  private hasError: boolean = false;

  @OnMutation({ childList: true, subtree: true })
  onMutation(): void {
    this.hasError = hasErrorSlot(this.root);
  }

  componentWillLoad(): void {
    this.setInput();
    this.setLabel();

    this.hasError = hasErrorSlot(this.root);

    trackComponent(this.root, { variant: this.variant });
  }

  disconnectedCallback(): void {
    this.disabledObserver.disconnect();
    this.requiredObserver.disconnect();
  }

  render(): JSX.Element {
    return (
      <GuxFormFieldContainer labelPosition={this.computedLabelPosition}>
        <GuxFormFieldLabel
          position={this.computedLabelPosition}
          required={this.required}
        >
          <slot name="label" onSlotchange={() => this.setLabel()} />
        </GuxFormFieldLabel>
        <div class="gux-input-and-error-container">
          <div
            class={{
              'gux-input': true,
              'gux-input-error': this.hasError
            }}
          >
            <div
              class={{
                'gux-input-container': true,
                'gux-disabled': this.disabled
              }}
            >
              <slot name="input" />
              <gux-icon icon-name="ic-dropdown-arrow" decorative></gux-icon>
            </div>
          </div>
          <GuxFormFieldError hasError={this.hasError}>
            <slot name="error" />
          </GuxFormFieldError>
        </div>
      </GuxFormFieldContainer>
    ) as JSX.Element;
  }

  private get variant(): string {
    const labelPositionVariant = this.labelPosition
      ? this.labelPosition.toLowerCase()
      : 'none';

    const type = this.input.getAttribute('type');

    return `${type}-${labelPositionVariant}`;
  }

  private setInput(): void {
    this.input = this.root.querySelector('select[slot="input"]');

    preventBrowserValidationStyling(this.input);

    this.disabled = this.input.disabled;
    this.required = this.input.required;

    this.disabledObserver = onDisabledChange(
      this.input,
      (disabled: boolean) => {
        this.disabled = disabled;
      }
    );
    this.requiredObserver = onRequiredChange(
      this.input,
      (required: boolean) => {
        this.required = required;
      }
    );

    validateFormIds(this.root, this.input);
  }

  private setLabel(): void {
    this.label = this.root.querySelector('label[slot="label"]');

    this.computedLabelPosition = getComputedLabelPosition(
      this.label,
      this.labelPosition
    );
  }
}
