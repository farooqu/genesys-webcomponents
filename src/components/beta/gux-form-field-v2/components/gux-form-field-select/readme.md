# gux-form-field-select-beta



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description | Type                                    | Default     |
| --------------- | ---------------- | ----------- | --------------------------------------- | ----------- |
| `labelPosition` | `label-position` |             | `"above" \| "beside" \| "screenreader"` | `undefined` |


## Slots

| Slot      | Description                     |
| --------- | ------------------------------- |
| `"error"` | Optional slot for error message |
| `"input"` | Required slot for input tag     |
| `"label"` | Required slot for label tag     |


## Dependencies

### Depends on

- [gux-icon](../../../../stable/gux-icon)

### Graph
```mermaid
graph TD;
  gux-form-field-select-beta --> gux-icon
  style gux-form-field-select-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
