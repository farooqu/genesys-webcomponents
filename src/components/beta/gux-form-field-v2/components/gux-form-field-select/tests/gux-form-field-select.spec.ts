import { newSpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxFormFieldSelect } from '../gux-form-field-select';

const components = [GuxFormFieldSelect];
const language = 'en';

describe('gux-form-field-select-beta', () => {
  beforeEach(async () => {
    (
      global as NodeJS.Global & {
        MutationObserver: any;
      }
    ).MutationObserver = MutationObserver;
  });

  it('should build', async () => {
    const html = `
      <gux-form-field-select-beta>
        <select slot="input" name="lp-1">
          <option value="option1">Option 1</option>
          <option value="option2">Option 2</option>
          <option value="option3">Option 3</option>
        </select>
        <label slot="label">Default</label>
      </gux-form-field-select-beta>
    `;
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxFormFieldSelect);
  });

  describe('#render', () => {
    describe('label-position', () => {
      [
        '',
        'label-position="above"',
        'label-position="beside"',
        'label-position="screenreader"'
      ].forEach((componentAttribute, index) => {
        it(`should render component as expected (${index + 1})`, async () => {
          const html = `
            <gux-form-field-select-beta ${componentAttribute}>
              <select slot="input" name="lp-1">
                <option value="option1">Option 1</option>
                <option value="option2">Option 2</option>
                <option value="option3">Option 3</option>
              </select>
              <label slot="label">Default</label>
            </gux-form-field-select-beta>
          `;
          const page = await newSpecPage({ components, html, language });

          expect(page.root).toMatchSnapshot();
        });
      });
    });

    describe('input attributes', () => {
      ['', 'disabled', 'required'].forEach((inputAttribute, index) => {
        it(`should render component as expected (${index + 1})`, async () => {
          const html = `
            <gux-form-field-select-beta>
              <select slot="input" name="lp-1" ${inputAttribute}>
                <option value="option1">Option 1</option>
                <option value="option2">Option 2</option>
                <option value="option3">Option 3</option>
              </select>
              <label slot="label">Default</label>
            </gux-form-field-select-beta>
          `;
          const page = await newSpecPage({ components, html, language });

          expect(page.root).toMatchSnapshot();
        });
      });
    });
  });
});
