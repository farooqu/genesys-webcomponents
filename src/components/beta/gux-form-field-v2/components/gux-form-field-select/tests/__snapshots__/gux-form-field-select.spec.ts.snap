// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-form-field-select-beta #render input attributes should render component as expected (1) 1`] = `
<gux-form-field-select-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select id="gux-form-field-input-i" name="lp-1" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;

exports[`gux-form-field-select-beta #render input attributes should render component as expected (2) 1`] = `
<gux-form-field-select-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select disabled="" id="gux-form-field-input-i" name="lp-1" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;

exports[`gux-form-field-select-beta #render input attributes should render component as expected (3) 1`] = `
<gux-form-field-select-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select id="gux-form-field-input-i" name="lp-1" required="" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;

exports[`gux-form-field-select-beta #render label-position should render component as expected (1) 1`] = `
<gux-form-field-select-beta>
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select id="gux-form-field-input-i" name="lp-1" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;

exports[`gux-form-field-select-beta #render label-position should render component as expected (2) 1`] = `
<gux-form-field-select-beta label-position="above">
  <mock:shadow-root>
    <div class="gux-above gux-form-field-container">
      <div class="gux-above gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select id="gux-form-field-input-i" name="lp-1" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;

exports[`gux-form-field-select-beta #render label-position should render component as expected (3) 1`] = `
<gux-form-field-select-beta label-position="beside">
  <mock:shadow-root>
    <div class="gux-beside gux-form-field-container">
      <div class="gux-beside gux-form-field-label">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select id="gux-form-field-input-i" name="lp-1" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;

exports[`gux-form-field-select-beta #render label-position should render component as expected (4) 1`] = `
<gux-form-field-select-beta label-position="screenreader">
  <mock:shadow-root>
    <div class="gux-form-field-container gux-screenreader">
      <div class="gux-form-field-label gux-screenreader">
        <slot name="label"></slot>
      </div>
      <div class="gux-input-and-error-container">
        <div class="gux-input">
          <div class="gux-input-container">
            <slot name="input"></slot>
            <gux-icon decorative="" icon-name="ic-dropdown-arrow"></gux-icon>
          </div>
        </div>
        <div class="gux-form-field-error">
          <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
          <div class="gux-message">
            <slot name="error"></slot>
          </div>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <select id="gux-form-field-input-i" name="lp-1" slot="input">
    <option value="option1">
      Option 1
    </option>
    <option value="option2">
      Option 2
    </option>
    <option value="option3">
      Option 3
    </option>
  </select>
  <label for="gux-form-field-input-i" slot="label">
    Default
  </label>
</gux-form-field-select-beta>
`;
